Add support for executable files without a `.sh` extension

rastersize opened the pull request:
As can be seen in the logs for 1473a6b (see build logs linked below), the current script doesn’t find the shell-script without the `.sh` extension. As can be seen in the following log after I added the file `tests/test-custom-no-extension` and set its file mode to `+x`:

```
Linting all executables and .sh files, ignoring files inside git modules...
[ OK ] Linting ./tests/ksh.sh...
[ OK ] Linting ./tests/sh.sh...
[ OK ] Linting ./tests/bash.sh...
[ OK ] Linting ./tests/test-custom-check.sh...
[ OK ] Linting ./build.sh...
[ OK ] Linting ./install.sh...
```

This is the same for both for the Linux and OS X builds:
- [Linux build log](https://travis-ci.org/rastersize/shell-ci-build/jobs/110600324#L155)
- [OS X build log](https://travis-ci.org/rastersize/shell-ci-build/jobs/110600325#L2441)

This PR resolves this issue by reverting PR #20’s change to the `find` command arguments and to use `+111` instead of `+x` for the `-perm` argument. Further it resolves compatibility with modern GNU `find` versions (4.5.12+, released in 2013) as it expects the value to be prepended by `/` instead.

I’ve also added a very rudimentary test that makes sure the built `find_cmd` fins the correct number of files.

@caarlos0 let me know if this looks good to you, or if there’s something you’d like me to modify.

rastersize committed:
Add test case for shell script without .sh

- It only has the permission bits set (chmod +x).

rastersize committed:
Revert "Merge pull request #20 from caarlos0/caarlos0-patch-1"

This reverts commit 10d764b396b3c564c77fc41fb2a6627d8aa07b30, reversing
changes made to 3ecb8b69053da48b25ef97612c80eef49e33dc8b.

rastersize committed:
Add support for modern GNU find

- GNU find v4.5.12 (released in 2013) and up have dropped support for
the `+` syntax. Instead `/` should be used before the mode mask.

rastersize committed:
Add rudimentary test that we find expected files

caarlos0 commented on build.sh:
```diff
@@ -28,7 +28,14 @@ find_prunes() {
+  # GNU `find` dropped compatibility support for the `+` syntax with 4.5.12 (release 2013), it
+  # instead uses `/` which has been supported for a long time. BSD `find` however still used `+`.
+  local perm_format_specifier="/"
+  if [ "$(uname -s)" = "Darwin" ]; then
```
I think we should test if the `find` called is the GNU find or BSD find, because I actually can have the GNU find in my path instead of the BSD's...

Maybe something like `find --version | grep GNU` will work good enough...

What do you think?

rastersize replied:
@caarlos0 Good point!
I’ll update the PR. It’ll be a little bit more involved since BSD find doesn’t accept `--version` option :disappointed:

caarlos0 commented:
@rastersize thanks for the PR, I have only one single comment..

rastersize committed:
Make GNU find detection much better

- Now looks specifically for GNU find instead of which OS we’re
currently being executed on.

rastersize commented:
@caarlos0 I’ve updated the PR with a commit that hopefully resolves your comment. Let me know what you think.

caarlos0 commented on build.sh:
```diff
@@ -28,7 +28,14 @@ find_prunes() {
+  # GNU `find` dropped compatibility support for the `+` syntax with 4.5.12 (release 2013), it
+  # instead uses `/` which has been supported for a long time. BSD `find` however still used `+`.
+  local perm_format_specifier="+"
+  if (( $(find --version /dev/null 2>&1 | grep GNU > /dev/null 2>&1; echo $?) == 0 )); then
```
since `find --version` on BSD returns an error, couldn't we just write this as  `if [ find --version ]; then GNU; else BSD; fi`?

rastersize committed:
Simplify detection of GNU find

rastersize commented:
Simplified it as suggested, a lot better now :smile: 
Still piping the output though, so that it doesn’t get printed in the log.

caarlos0 commented:
Nice, thanks!
